### ISSUE TITLE
false-positive(reportArgumentType):  `"__hash__" is not present`

### ISSUE THREAD

**Issue by lucascolley** [DetachHead/basedpyright#1779]
### Description

This program works as expected at runtime:

```console
❯ pixi exec -w numpy ipython
```

```python
from collections.abc import Hashable

import numpy as np


def f(g: Hashable) -> int:
    return hash(g)

assert hasattr(np, "__hash__")
assert isinstance(hash(np), int)
assert f(np) == hash(np)
```

but basedpyright reports:

```
Argument of type "Module("numpy")" cannot be assigned to parameter "g" of type "Hashable" in function "f"
  "__hash__" is not present (basedpyright reportArgumentType)
```

on the `f(np)` call.

---

I couldn't find a duplicate issue here or upstream. I haven't checked whether this problem exists in `pyright` too.

**Comment by KotlinIsland** [comment]
very interesting case, I can investigate soon

**Comment by DetachHead** [comment]
minimal repro:

```py
import types

types.__hash__ # error: "__hash__" is not a known attribute of module "types" 
```
since it's present on `object` technically everything should be considered hashable, but since `__hash__` is implemented in the stupidest way imaginable, it's not typesafe at all (ie. overrides can set it to `None`)

while playing around with these dunders i also discovered #1781. perhaps we can fix by doing this:

```py
this._addImplicitSymbolToCurrentScope('__hash__', node, 'Callable[[], int] | None');
```
